### ISSUE TITLE
Refactoring password.reset

### ISSUE THREAD

**Issue by stuartpb** [opensets/domainprofiles#70]
Looking at #4, #37, and #38, I think it might make more sense to just refactor many of the fields of password.reset to something more like `password.reset.steps`, which would be an array detailing what happens when a reset token is activated:
- change
- autochange (for when the password is changed to something else by the site, a la Neocities)
- button (when the next step happens automatically, but only after a POST)
- ~~expire (documenting at what point the token is expired)~~ [see comment below](https://github.com/opensets/domainprofiles/issues/70#issuecomment-213896145)
- stub (the page displays a message confirming the reset and you go nowhere)
- login (user is directed to the login page)
- autologin (user is logged in)

So, the best sites go `[change, expire, autologin]`, Bountysource goes `[expire, change, autologin]`, Neocities goes `[autochange]` (and not even `[autochange, login]`), and there are lots of weird different ways this can go down that are all worth documenting.

**Comment by stuartpb** [comment]
Also, with this change, `password.reset.response.token` should really just become `password.reset.token`.

**Comment by stuartpb** [comment]
This is such an elegant solution to a number of fairly significant problems with the current schema that I'm marking it as a v0.1.0 blocker.

**Comment by stuartpb** [comment]
Profiles that need to be rewritten to support password.reset.steps:

- [ ] [37signals.com](https://github.com/opensets/domainprofiles/blob/master/profiles/37signals.com.yaml)
- [ ] [adobe.com](https://github.com/opensets/domainprofiles/blob/master/profiles/adobe.com.yaml)
- [ ] [bountysource.com](https://github.com/opensets/domainprofiles/blob/master/profiles/bountysource.com.yaml)
- [ ] [bugs.gnupg.org](https://github.com/opensets/domainprofiles/blob/master/profiles/bugs.gnupg.org.yaml)
- [ ] [compose.io](https://github.com/opensets/domainprofiles/blob/master/profiles/compose.io.yaml)
- [ ] [digital.darkhorse.com](https://github.com/opensets/domainprofiles/blob/master/profiles/digital.darkhorse.com.yaml)
- [ ] [dynadot.com](https://github.com/opensets/domainprofiles/blob/master/profiles/dynadot.com.yaml)
- [ ] [github.com](https://github.com/opensets/domainprofiles/blob/master/profiles/github.com.yaml)
- [ ] [godaddy.com](https://github.com/opensets/domainprofiles/blob/master/profiles/godaddy.com.yaml)
- [ ] [guru.com](https://github.com/opensets/domainprofiles/blob/master/profiles/guru.com.yaml)
- [ ] [hub.docker.com](https://github.com/opensets/domainprofiles/blob/master/profiles/hub.docker.com.yaml)
- [ ] [mailchimp.com](https://github.com/opensets/domainprofiles/blob/master/profiles/mailchimp.com.yaml)
- [ ] [name.com](https://github.com/opensets/domainprofiles/blob/master/profiles/name.com.yaml)
- [ ] [neocities.org](https://github.com/opensets/domainprofiles/blob/master/profiles/neocities.org.yaml)
- [ ] [odesk.com](https://github.com/opensets/domainprofiles/blob/master/profiles/odesk.com.yaml)
- [ ] [persona.org](https://github.com/opensets/domainprofiles/blob/master/profiles/persona.org.yaml)
- [ ] [qfc.com](https://github.com/opensets/domainprofiles/blob/master/profiles/qfc.com.yaml)
- [ ] [readthedocs.org](https://github.com/opensets/domainprofiles/blob/master/profiles/readthedocs.org.yaml)
- [ ] [slides.com](https://github.com/opensets/domainprofiles/blob/master/profiles/slides.com.yaml)
- [ ] [wikimedia.org](https://github.com/opensets/domainprofiles/blob/master/profiles/wikimedia.org.yaml)

**Comment by stuartpb** [comment]
`steps` could possibly include where you go after submitting the reset form - IFTTT takes you to the login page.

**Comment by stuartpb** [comment]
#52 raises an interesting wrinkle: what about sites where visiting the page *starts an accelerated timer* for the token's expiration?

This... should probably remain a separate field, as proposed in #52, which doesn't interact with `steps`.

**Comment by stuartpb** [comment]
`expire` steps are now logged by recording what step they coincide with as `password.reset.expiration.trigger`.

**Comment by stuartpb** [comment]
Actually, how should I record MediaWiki's password reset? It's like "log in with reset password -> change password -> automatically logged in", so would the steps be `[login, change, autologin]`?

**Comment by stuartpb** [comment]
As I just alluded to in #104, I'm considering expanding this to include `request` and `visit` steps explicitly, to support a few sites with more complex schemes, although I don't think that's the right way to advance just yet.

**Comment by stuartpb** [comment]
orcacard.com requires you to answer a security question before it changes your password to a new one and emails it to you.

**Comment by stuartpb** [comment]
Considering that FluxBB (as used by bbs.archlinux.org) does the "visit a link to change your password to this random one we're sending you" thing, I think something about this needs to be adjusted - like explicitly noting the `request` and `visit` steps, or augmenting with an `autochange` field describing how arbitrarily-set passwords get set (this should probably be suggested as a new issue).

**Comment by stuartpb** [comment]
Maybe a `generate` step (for when a password is randomly-generated)?

**Comment by stuartpb** [comment]
I think it needs to be more specific than `generate`, to distinguish it from non-password generation (ie.tokens).

**Comment by stuartpb** [comment]
I think the step where a password is changed to be random should be `randomize`, and the password disclosure should be in `response` (which should probably be `responses`).

**Comment by stuartpb** [comment]
Better than `responses`, I think `response.body` should get replaced with `response.messages`, which is an array of (usually one) strings of the form of what `body` is now - and each message corresponds to a sequential `response` step in `steps`.

Of course, this assumes each response comes from the same sender, but I think that's a safe assumption to make - any system where that *isn't* the case probably isn't using consistent enough addresses to even record a `password.reset.response.sender` address in the first place.

**Comment by stuartpb** [comment]
On the other hand- and I think this is kind of clearly where this has needed to go for a while now - I think the most sensible design for `password.reset.steps` is to have *each step be an object*, and have these steps simply assimilate other fields on `password.reset` like `request` and `response`.

**Comment by stuartpb** [comment]
Per #127, domains using an approach drastically different than traditional password flow going forward will go under a completely different set of keys, so `steps` doesn't have to reflect those distinctions any more.

**Comment by stuartpb** [comment]
Honestly, with the advent of #127's approach, I'm thinking of rolling this "steps" object *back*, with new fields to better describe the factors it's cataloguing: expiration is already recorded with `password.reset.flow.expiration`, and submission behaviors could better be described with something like `password.reset.flow.submit.sessions` (which should really be refactored: I'm going to describe that plan over there).

Closing in favor of a unified #127.
